src/defrag.c

```diff
@@ -840,299 +914,450 @@ int defragLaterItem(dictEntry *de, unsigned long *cursor, long long endtime, int
-                server.active_defrag_running = 0;
-
-                computeDefragCycles(); /* if another scan is needed, start it right away */
-                if (server.active_defrag_running != 0 && ustime() < endtime) continue;
-                break;
-            } else if (current_db == 0) {
-                /* Start a scan from the first database. */
-                start_scan = ustime();
-                start_stat = server.stat_active_defrag_hits;
-            }
+    server.stat_total_active_defrag_time += elapsedUs(server.stat_last_active_defrag_time);
+    server.stat_last_active_defrag_time = 0;
+    server.active_defrag_cpu_percent = 0;
+}
+
+
+/* Must be called at the start of the timeProc as it measures the delay from the end of the previous
+ * timeProc invocation when performing the computation. */
+static int computeDefragCycleUs(void) {
+    long dutyCycleUs;
+
+    int targetCpuPercent = server.active_defrag_cpu_percent;
+    serverAssert(targetCpuPercent > 0 && targetCpuPercent < 100);
 
-            db = &server.db[current_db];
-            kvstoreDictLUTDefrag(db->keys, dictDefragTables);
-            kvstoreDictLUTDefrag(db->expires, dictDefragTables);
-            defrag_stage = 0;
-            defrag_cursor = 0;
-            slot = -1;
-            defrag_later_item_in_progress = 0;
+    static int prevCpuPercent = 0; // STATIC - this persists
+    if (targetCpuPercent != prevCpuPercent) {
+        /* If the targetCpuPercent changes, the value might be different from when the last wait
+         *  time was computed.  In this case, don't consider wait time.  (This is really only an
+         *  issue in crazy tests that dramatically increase CPU while defrag is running.) */
+        defrag.timeproc_end_time = 0;
+        prevCpuPercent = targetCpuPercent;
+    }
+
+    // Given when the last duty cycle ended, compute time needed to achieve the desired percentage.
+    if (defrag.timeproc_end_time == 0) {
+        // Either the first call to the timeProc, or we were paused for some reason.
+        defrag.timeproc_overage_us = 0;
+        dutyCycleUs = server.active_defrag_cycle_us;
+    } else {
+        long waitedUs = getMonotonicUs() - defrag.timeproc_end_time;
+        /* Given the elapsed wait time between calls, compute the necessary duty time needed to
+         *  achieve the desired CPU percentage.
+         *  With:  D = duty time, W = wait time, P = percent
+         *  Solve:    D          P
+         *          -----   =  -----
+         *          D + W       100
+         *  Solving for D:
+         *     D = P * W / (100 - P)
+         *
+         * Note that dutyCycleUs addresses starvation.  If the wait time was long, we will compensate
+         *  with a proportionately long duty-cycle.  This won't significantly affect perceived
+         *  latency, because clients are already being impacted by the long cycle time which caused
+         *  the starvation of the timer. */
```

Comment:
While it is correct, this is not a good behaviour, especially from tail latency perspective. A better behaviour would increase the "measure" time interval to keep smooth run time of defrag even when there exists a periodic spike. Now a sporadic long running command\cron-job will cause defrag to run in sporadic long running spikes as well, isn't it so?

Reply:
dutyCycleUs is the upper bound, not the lower bound. We bias against latency spikes, especially when the system is busy.

Reply:
@madolson that's not quite right.  If the timer is getting starved, we will increase the duty-cycle to combat the starvation.  This won't occur unless there is a delay to the timer.

@zvi, I don't agree with this being bad behavior though.  And the tail latency is unimportant in this case.  Defrag is possibly the most critical maintenance task in the system.  With today's defrag, `serverCron` is subject to starvation.  A customer generating a series of slow commands will severely impede defrag.  There is an attempt at anti-starvation code  (which I removed) in `whileBlockedCron`.  Reference: https://github.com/valkey-io/valkey/pull/1242/files#diff-1abc5651133d108c0c420d9411925373c711133e7748d9e4f4c97d5fb543fdd9L1617-L1626
(Note - in re-reviewing that code, I think we still need a hook there - as timers aren't run while blocked.  I'll add a separate comment on that.)

Consider a command that takes 2 seconds to execute - and delay's all of the timers.  In order to get 10% of the CPU (for example), we need to run for 200ms to compensate for the starvation.  HOWEVER, that 200ms is small as compared to the 2-seconds that the original command took.  Tail latencies are dominated by the slow command, and not by active defrag.  Likewise, if the event loop is slowed due to overload, the tail latencies are dominated by the overload, not the additional fraction for defrag.

Reply:
@JimB123 , I think we should implement a simple mechanism that would smooth usage peaks when possible, but it's not a blocker for this PR and can be suggested in followup improvement.